File: src/code-samples/langgraph/langgraph-functional-api-stream-custom-data.py
from langchain_core.utils.uuid import uuid7
from langgraph.checkpoint.memory import InMemorySaver
from langgraph.config import get_stream_writer
from langgraph.func import entrypoint

checkpointer = InMemorySaver()


@entrypoint(checkpointer=checkpointer)
def main(inputs: dict) -> int:
    writer = get_stream_writer()
    writer("Started processing")
    result = inputs["x"] * 2
    writer(f"Result is {result}")
    return result


# :snippet-start: langgraph-functional-api-stream-custom-data-py
config = {"configurable": {"thread_id": str(uuid7())}}

stream = main.stream_events({"x": 5}, config=config, version="v3")
for mode, chunk in stream.interleave("values"):
    print(f"{mode}: {chunk}")
# values: 10
# :snippet-end:


# :remove-start:
test_config = {"configurable": {"thread_id": str(uuid7())}}
test_stream = main.stream_events({"x": 5}, config=test_config, version="v3")
chunks = list(test_stream.interleave("values"))

assert chunks == [
    ("values", 10),
]
print("✓ langgraph-functional-api-stream-custom-data")
# :remove-end:

File: src/code-samples/langgraph/langgraph-functional-api-stream-custom-data.ts
import { entrypoint, MemorySaver } from "@langchain/langgraph";
import type { LangGraphRunnableConfig } from "@langchain/langgraph";

const checkpointer = new MemorySaver();

const main = entrypoint(
  { checkpointer, name: "main" },
  async (
    inputs: { x: number },
    config: LangGraphRunnableConfig,
  ): Promise<number> => {
    config.writer?.("Started processing");
    const result = inputs.x * 2;
    config.writer?.(`Result is ${result}`);
    return result;
  },
);

// :snippet-start: langgraph-functional-api-stream-custom-data-js
const config = {
  configurable: { thread_id: "functional-api-stream-custom-data" },
};

const stream = await main.streamEvents({ x: 5 }, { ...config, version: "v3" });
for await (const chunk of stream.values) {
  console.log(chunk);
}
// 10
// :snippet-end:

// :remove-start:
const testStream = await main.streamEvents(
  { x: 5 },
  { ...config, version: "v3" },
);
const chunks = [];
for await (const chunk of testStream.values) {
  chunks.push(chunk);
}

if (JSON.stringify(chunks) !== JSON.stringify([10])) {
  throw new Error(`Expected [10], got ${JSON.stringify(chunks)}`);
}
console.log("✓ langgraph-functional-api-stream-custom-data-js");
// :remove-end:
